### PR TITLE
BTree: add drain methods

### PR DESCRIPTION
**Pull request by ssomers**
Implements #81074.
Builds on #93989.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Is there a chance we can split this up into more commits? Reviewing a diff in a couple thousand lines is going to be quite challenging. It would be good to document any design decisions or otherwise guide my review, too.

**Comment by ssomers** [comment]
I can split off a few refactorings of existing functions, to open up for new use, but they seemed a bit pointless on their own (apart from one thing that I actually postponed).

**Comment by Mark-Simulacrum** [comment]
I don't really know what you postponed, but I will likely not be able to effectively review this diff without at least some guidance as to where to start or the overall vision in adding these methods.

**Comment by ssomers** [comment]
The item I postponed is that `steal_X` functions (apart from their return value) are basically a hand-written optimization of the generic `bulk_steal_X` functions for count=1 (in reality, it's the other way around of course). It comes up here because one of the functions needed for `drain` is identical to `handle_shrunk_node_recursively` apart from that alleged optimization.

**Comment by ssomers** [comment]
The overall vision in adding these methods is to not copy any existing functionality, but break it up in order to reuse parts. I'll feed those refactings in a few separate PRs now.

**Comment by ssomers** [comment]
Added panics-clause to function docs

**Comment by jeffparsons** [comment]
If you'll take comments from the peanut gallery...

This series of work by ssomers (et al.) is one of the most exciting standard library cleanups/additions in quite a while for me. In a perfect world we'll eventually have a full mutable cursor API, but even this range-based drain will unlock a lot of extra performance and simplicity for me, especially in my [rangemap](https://github.com/jeffparsons/rangemap) crate.

Many thanks for all the hard work that's gone in here! 💖

**Comment by ssomers** [comment]
Thanks, it's nice to know there was a point to it.

**Comment by JohnCSimon** [comment]
Ping from triage:
@ssomers Can you please resolve the merge conflict?

**Comment by ssomers** [comment]
No problem. Apparently github no longer notifies of such conflicts.

File: library/alloc/src/collections/btree/map.rs [hunk 1]
```diff
@@ -1626,6 +1737,36 @@ impl<K, V> Clone for Values<'_, K, V> {
...
+        self.inner.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<K, V> ExactSizeIterator for Drain<K, V> {
```
Comment (the8472) [inline library/alloc/src/collections/btree/map.rs:1761]: If it is exact, is it also trusted?

Comment (ssomers) [reply library/alloc/src/collections/btree/map.rs:1761]: I trust it… but you probably mean #37306. If I understand correctly, every [ExactSizeIterator](https://doc.rust-lang.org/core/iter/trait.ExactSizeIterator.html) sourcing from a collection (so not adapting like Skip) should be [TrustedLen](https://doc.rust-lang.org/core/iter/trait.TrustedLen.html). But in BTree, none is, so I probably don't.

Comment (the8472) [reply library/alloc/src/collections/btree/map.rs:1761]: > but you probably mean #37306.

Yes.

> But in BTree, none is, so I probably don't.

So is there any reason not to trust it, can it be wrong?

Comment (ssomers) [reply library/alloc/src/collections/btree/map.rs:1761]: The `Drain` iteraror is a thin wrapper aroind `IntoIter`, albeit instantiated on a map that may well violate some invariants, but only invariants irrelevant for iteration. So the same reasons for distrust hold as why `IntoIter` is not trusted. I have no idea what those reasons could be, for `IntoIter`, for `Drain`, or for any simple iterator over a collection. I imagine it's legal and might even be beneficial to declare a `DrainFilter` iterator as `ExactSizeIterator`, stating it 's intende to be used when the predicate mostly yields true (which is the opposite of the current reality), and then that iterator shouldn't be `TrustedLen` at all. But that's an entirely different situation from `Drain`.

Comment (the8472) [reply library/alloc/src/collections/btree/map.rs:1761]: > I imagine it's legal and might even be beneficial to declare a DrainFilter iterator as ExactSizeIterator

From a safety perspective it is. From an API contract perspective it is not.

> When doing so, the implementation of Iterator::size_hint must return the exact size of the iterator.

**Comment by apiraino** [comment]
Are the labels `T-compiler`+`T-rustdoc` relevant here or is it `T-libs` only?

**Comment by ssomers** [comment]
I don't really know what those labels mean or how bors decides to slap them onto a PR, but I strongly doubt it got it right.

**Comment by apiraino** [comment]
`T-*` labels indicate which team is relevant for an issue or a pull request and help organize the work. In this case I think bors was a bit overzealous :)

@rustbot label -T-rustdoc -T-compiler
Assigning priority as discussed in the [Zulip thread](#) of the Prioritization Working Group.

**Comment by ssomers** [comment]
To stir up things, yanked the `drain` interface. I added `remove_range` instead just to have something that drops key-value pairs. Which only now makes me realize the `drop_panic_leak` test case for `split_off_range` is stupid: nothing is dropped before the split is finished.

**Comment by yaahc** [comment]
r? rust-lang/libs-api

**Comment by joshtriplett** [comment]
r? rust-lang/libs-api

**Pull request closed without merging**